source/auth/auth.rst

```diff
@@ -1159,6 +1159,274 @@ If AWS authentication fails for any reason, the cache MUST be cleared.
+`````````````````````````````
+
+username
+    MUST NOT be specified.
+
+source
+    MUST be "$external". Defaults to ``$external``.
+
+password
+    MUST NOT be specified.
+
+mechanism
+    MUST be "MONGODB-OIDC"
+
+mechanism_properties
+    PRINCIPAL_NAME
+        Drivers MUST allow the user to specify a principal name, which
+        is required by the server when more than one Identity Provider
+        is configured.
+    DEVICE_NAME
+        Drivers MUST allow the user to specify a name for the device
+        workflow that is one of "aws", "azure", or "gcp".
+    REQUEST_TOKEN_CALLBACK
+        Drivers MUST allow the user to specify a callback of the form
+        "onOIDCRequestToken" (defined below), if the driver supports
+        providing objects as mechanism property values.
+    REFRESH_TOKEN_CALLBACK
+        Drivers MUST allow the user to specify a callback of the form
+        "onOIDCRefreshToken" (defined below), if the driver supports
+        providing objects as mechanism property values.
+
+Drivers MUST skip client step 1 for device workflows
+or when the server step 1 response value is cached.  When skipping step 1,
+drivers will use ``saslStart`` and a payload with the ``jwt`` value.
+
+User Provided Callbacks
+```````````````````````
+
+Drivers MUST allow the user to provide callbacks for token request and
+token refresh.  The token request callback MUST accept the OIDCMechanismServerStep1 structure and return an OIDCRequestTokenResult of
+the form:
+
+.. code:: idl
+
+  OIDCRequestTokenResult:
+      description: "The result of a token request"
+      strict: false
+      fields:
+          accessToken:
+              description: "The OIDC access token"
+              type: string
+          expiresInSeconds:
+              description: "The expiration time in seconds from the current time"
+              type: int
+              optional: true
+          refreshToken:
+              description: "The OIDC refresh token"
+              type: str
+              optional: true
+
+.. code:: typescript
+
+    function onOIDCRequestToken(serverInfo: OIDCMechanismServerStep1): OIDCRequestTokenResult
+
+Callbacks can be synchronous and/or asynchronous, depending on the driver
+and/or language.  Asynchronous callbacks should be preferred when other
+operations in the driver use asynchronous functions.
+
+Before calling a callback, the driver MUST acquire a lock, and verify that
+there are no valid cache credentials before calling the callback.
+This is because request callbacks may involve human interaction, and refresh
+callbacks could use refresh tokens that can only be used once.
+
+The driver MUST provide a way for the callback to be either automatically
+cancelled, or to cancel itself.  This can be as a timeout argument to the
+callback, a cancellation context passed to the callback, or some other
+language-appropriate mechanism.  The timeout MUST respect the client side
+operation timeout if set, or 60 seconds.
```

Comment:
Does this mean that there is a hard limit of 60 seconds or CSOT for completing authentication? That seems a bit restrictive for human flows (think 2FA codes taking a while to propagate etc.)

Reply:
I agree.  The user might also have to manually retrieve a password from a vault, etc.  The question is then whether we want to impose any time limit on this.  Perhaps 5 minutes?

Reply:
I’d expect any time limit to be configured by the callback itself. 5 minutes seems like it should be enough for human flows and I’d definitely be okay with using that here, I’m just wondering if the driver is really the right place to enforce a limit.

Reply:
I would say differently. On the driver side, are we ok with possibly infinite waiting under lock? Since user interaction can potentially take any time. I would say not relly

Reply:
I agree, even if the driver is multi-threaded, we still do not want to have an orphaned thread.  I've updated this section.